File: 1998/1/5-1.py
#!/usr/bin/python
class Node:
	def __init__(self, v = " "):
		self.val = v
		self.child = []

def match(name = "", n = 1, names = [], bill = []):
	name += bill[n - 1].val
	if bill[n - 1].child == []:
		names.append(name)
	else:
		for i in xrange(len(bill[n - 1].child)):
			match(name, bill[n - 1].child[i], names, bill)

if __name__ == "__main__":
	fin = file("prob1.in", "r")
	cnt = 1
	while 1:
		line = fin.readline()
		if line == "":
			break
		if line[-1] == "\n":
			line = line[:-1]
		names = []
		num, m, v = line.split()
		num = int(num)
		m = int(m)
		bill = [None] * num
		bill[0] = Node(v)
		for i in xrange(m):
			line = fin.readline()
			p, num, s, v = line.split()
			p = int(p)
			num = int(num)
			bill[num - 1] = Node(v)
			bill[p - 1].child.append(num)
		match("", 1, names, bill)
		full = 1
		part = 0
		for name in names:
			if name != names[0]:
				full = 0
			if names.count(name) > 1:
				part = 1
		print "Sample %s:" % cnt,
		if full == 1:
			print "Exact match - summon the grand jury."
		elif part == 1:
			print "Partial match - close, but no cigar."
		else:
			print "No match - sorry, Ken."
		cnt += 1
		

File: 1998/1/5-1.rb
#!/usr/bin/env ruby

class Node
  attr_reader :value

  def initialize(v = " ")
    @value = v
    @child = []
  end

  def empty?
    @child.empty?
  end

  def <<(x)
    @child << x
  end

  def length
    @child.length
  end

  def [](i)
    @child[i]
  end
end

def match(name = "", n = 1, names = [], bill = [])
  name = name + bill[n-1].value
  if bill[n-1].empty?
    names << name
  else
    bill[n-1].length.times do |i|
      match(name, bill[n-1][i], names, bill)
    end
  end
end

count = 1
until STDIN.eof do
  line = STDIN.readline
  break if line.empty?
  names = []
  num, m, v = line.split
  num = num.to_i
  m = m.to_i
  bill = [nil] * num
  bill[0] = Node.new(v)
  m.times do |i|
    p, num, s, v = STDIN.readline.split
    p = p.to_i
    num = num.to_i
    bill[num - 1] = Node.new(v)
    bill[p - 1] << num
  end
  match("", 1, names, bill)
  full = 1
  part = 0
  names.each do |name|
    full = 0 if name != names[0]
    part = 1 if names.count(name) > 1
  end
  print "Sample #{count}: "
  if full == 1
    puts "Exact match - summon the grand jury."
  elsif part == 1
    puts "Partial match - close, but no cigar."
  else
    puts "No match - sorry, Ken."
  end
  count += 1
end
